File: QuintaSemana/pontes_magicas.py
# 1706 - Pontes Mágicas


from collections import deque
import sys


def bfs_teste(grafo, no_inicial, sons, visitados):
    contador = 0

    q = deque([no_inicial])
    while q:
        no = q.popleft()
        if visitados[no-1] == 0:
            visitados[no-1] = 1
            if sons[no-1] == 'B':
                contador += 1
            for vizinho in grafo[no]:
                if visitados[vizinho-1] == 0:
                    q.append(vizinho) 
                    
    #print(f'ligacoes: {qtd_ligacoes_BB//2}')
    return contador

sys.stdin = open('teste.txt', 'r')

def main():
    resultados = []
    while True:
        try:
            qtd_torres, m = map(int, input().split())
            sons_iniciais = input().split()
            conexoes = {}
            for _ in range(m):
                a, b = map(int, input().split())
                conexoes.setdefault(a, []).append(b)
                conexoes.setdefault(b, []).append(a)

            lista_visitados = [0] * qtd_torres
            #print(f'VISITADOS: {lista_visitados}')
            possivel = True

            #print(f'Grafo: {conexoes}')
            #print(f'Sons: {sons_iniciais}')
            for i in range(qtd_torres):
                if lista_visitados[i] == 0:
                    contador = bfs_teste(conexoes,i+1,sons_iniciais,lista_visitados)
                    #print(f'CONTADOR: {contador}')
                    if contador % 2 != 0:
                        possivel = False
                        break    
            if possivel:
                resultados.append('Y')
            else:
                resultados.append('N')

        except EOFError:
            break
    print('\n'.join(resultados))



if __name__ == '__main__':
    main()              
                    

g = {
    1: [2],
    2: [1],
}



# print(dfs(g,1,['A', 'B']))

def dfs(grafo, no_inicial, sons):
    visitados = []
    qtd_ligacoes_BB = 0 

    pilha = [no_inicial]
    while pilha:
        no = pilha.pop()
        if no not in visitados:
            visitados.append(no)
            pilha.extend(reversed(grafo[no]))
            for v in grafo[no]:
                if sons[no-1] == 'B' and sons[v-1] == 'B':
                    #print('achei a aresta B-B')
                    qtd_ligacoes_BB += 1
                    
    #print(f'ligacoes: {qtd_ligacoes_BB//2}')
    return (qtd_ligacoes_BB//2) % 2 == 0

File: QuintaSemana/pontes_magicas.cpp
#include <iostream>
#include <vector>
#include <queue>
#include <map>

using namespace std;

int bfs_teste(map<int, vector<int>>& grafo, int no_inicial, vector<char>& sons, vector<int>& visitados) {
    int contador = 0;
    queue<int> q;
    q.push(no_inicial);
    
    while (!q.empty()) {
        int no = q.front();
        q.pop();
        
        if (visitados[no] == 0) {
            visitados[no] = 1;
            if (sons[no] == 'B') {
                contador++;
            }
            for (int vizinho : grafo[no]) {
                if (visitados[vizinho] == 0) {
                    q.push(vizinho);
                }
            }
        }
    }
    return contador;
}

int main() {
    vector<char> resultados;
    
    while (true) {
        int qtd_torres, m;
        if (!(cin >> qtd_torres >> m)) break;
        
        vector<char> sons_iniciais(qtd_torres);
        for (int i = 0; i < qtd_torres; i++) {
            cin >> sons_iniciais[i];
        }
        
        map<int, vector<int>> conexoes;
        for (int i = 0; i < m; i++) {
            int a, b;
            cin >> a >> b;
            conexoes[a - 1].push_back(b - 1);
            conexoes[b - 1].push_back(a - 1);
        }
        
        vector<int> lista_visitados(qtd_torres, 0);
        bool possivel = true;
        
        for (int i = 0; i < qtd_torres; i++) {
            if (lista_visitados[i] == 0) {
                int contador = bfs_teste(conexoes, i, sons_iniciais, lista_visitados);
                if (contador % 2 != 0) {
                    possivel = false;
                    break;
                }
            }
        }
        
        resultados.push_back(possivel ? 'Y' : 'N');
    }
    
    for (char res : resultados) {
        cout << res << '\n';
    }
    
    return 0;
}